File: apps/api/src/utils/classes.ts
import { getLevelsForSchoolType } from "@makyschool/shared/constants";
import type { SchoolType } from "@makyschool/shared/types";
import { pool } from "../db/pool.js";

export function formatClassLabel(level: string, stream: string | null): string {
  return `${level}${stream ?? ""}`;
}

export function isLevelAllowedForSchoolType(level: string, schoolType: SchoolType | null): boolean {
  return getLevelsForSchoolType(schoolType).includes(level);
}

export async function getSchoolType(schoolId: string): Promise<SchoolType | null> {
  const result = await pool.query<{ school_type: SchoolType | null }>(
    "SELECT school_type FROM schools WHERE id = $1",
    [schoolId],
  );
  return result.rows[0]?.school_type ?? null;
}

export function getAllowedLevelsSqlParam(schoolType: SchoolType | null): string[] {
  return getLevelsForSchoolType(schoolType);
}

export function buildLevelOrderCase(columnRef: string, schoolType: SchoolType | null): string {
  const levels = getLevelsForSchoolType(schoolType);
  if (levels.length === 0) {
    return "0";
  }

  const cases = levels
    .map((level, index) => `WHEN ${columnRef} = '${level}' THEN ${index}`)
    .join(" ");

  return `CASE ${cases} ELSE ${levels.length} END`;
}

export async function findDuplicateClass(
  schoolId: string,
  level: string,
  stream: string | null,
  excludeId?: string,
): Promise<boolean> {
  const result = await pool.query(
    `SELECT id FROM school_classes
     WHERE school_id = $1
       AND level = $2
       AND COALESCE(stream, '') = COALESCE($3, '')
       AND ($4::uuid IS NULL OR id <> $4)
     LIMIT 1`,
    [schoolId, level, stream ?? "", excludeId ?? null],
  );
  return (result.rowCount ?? 0) > 0;
}

File: apps/api/app/lib/classes.py
import uuid
from typing import Literal

import asyncpg

SchoolType = Literal["primary", "secondary", "both"]

PRIMARY_CLASS_LEVELS = ("P1", "P2", "P3", "P4", "P5", "P6", "P7")
SECONDARY_CLASS_LEVELS = ("S1", "S2", "S3", "S4", "S5", "S6")


def get_levels_for_school_type(school_type: SchoolType | str | None) -> list[str]:
    if school_type == "secondary":
        return list(SECONDARY_CLASS_LEVELS)
    if school_type == "both":
        return list(PRIMARY_CLASS_LEVELS) + list(SECONDARY_CLASS_LEVELS)
    return list(PRIMARY_CLASS_LEVELS)


def format_class_label(level: str, stream: str | None) -> str:
    return f"{level}{stream or ''}"


def is_level_allowed_for_school_type(level: str, school_type: SchoolType | str | None) -> bool:
    return level in get_levels_for_school_type(school_type)


async def get_school_type(conn: asyncpg.Connection, school_id: uuid.UUID) -> SchoolType | str | None:
    row = await conn.fetchrow(
        "SELECT school_type FROM schools WHERE id = $1",
        school_id,
    )
    return row["school_type"] if row else None


def get_allowed_levels_sql_param(school_type: SchoolType | str | None) -> list[str]:
    return get_levels_for_school_type(school_type)


def build_level_order_case(column_ref: str, school_type: SchoolType | str | None) -> str:
    levels = get_levels_for_school_type(school_type)
    if not levels:
        return "0"

    cases = " ".join(
        f"WHEN {column_ref} = '{level}' THEN {index}" for index, level in enumerate(levels)
    )
    return f"CASE {cases} ELSE {len(levels)} END"


async def find_duplicate_class(
    conn: asyncpg.Connection,
    school_id: uuid.UUID,
    level: str,
    stream: str | None,
    exclude_id: uuid.UUID | None = None,
) -> bool:
    row = await conn.fetchrow(
        """
        SELECT id FROM school_classes
        WHERE school_id = $1
          AND level = $2
          AND COALESCE(stream, '') = COALESCE($3, '')
          AND ($4::uuid IS NULL OR id <> $4)
        LIMIT 1
        """,
        school_id,
        level,
        stream or "",
        exclude_id,
    )
    return row is not None
